pkg/cdi/producer/producer.go

```diff
@@ -0,0 +1,188 @@
+/*
+   Copyright © 2024 The CDI Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package producer
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"sigs.k8s.io/yaml"
+
+	"tags.cncf.io/container-device-interface/pkg/cdi/producer/validator"
+	cdi "tags.cncf.io/container-device-interface/specs-go"
+)
+
+// A SpecProducer defines a structure for outputting CDI specifications.
+type SpecProducer struct {
+	*cdi.Spec
+	options
+}
+
+// New creates a new producer with the supplied options.
+func New(raw *cdi.Spec, opts ...Option) (*SpecProducer, error) {
+	sp := &SpecProducer{
+		Spec: raw,
+		options: options{
+			overwrite: true,
+			// TODO: This could be updated to 0644 to be world-readable.
+			permissions:   0600,
+			specFormat:    DefaultSpecFormat,
+			specValidator: validator.Default,
+		},
+	}
+	return sp.applyOptions(opts...)
+}
+
+// cloneWithOptions creates a copy of the specified spec producer with the given options.
+func cloneWithOptions(p *SpecProducer, opts ...Option) (*SpecProducer, error) {
+	sp := &SpecProducer{
+		Spec:    p.Spec,
+		options: p.options,
+	}
+	return sp.applyOptions(opts...)
+}
+
+// applyOptions applies the specified options to a SpecProducer
+func (p *SpecProducer) applyOptions(opts ...Option) (*SpecProducer, error) {
+	for _, opt := range opts {
+		err := opt(&p.options)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return p, nil
+}
+
+// Save writes a CDI spec to a file with the specified name.
+// If the filename ends in a supported extension, the format implied by the
+// extension takes precedence over the format with which the SpecProducer was
+// configured.
+func (p *SpecProducer) Save(filename string) (string, error) {
+	filename = p.normalizeFilename(filename)
+	format := p.specFormatFromFilename(filename)
+	// If the currently configured format doesn't match the expected format for
+	// the filename, we create a new
+	if p.specFormat != format {
+		pWithFormat, err := cloneWithOptions(p, WithSpecFormat(format))
+		if err != nil {
+			return "", err
+		}
+		return pWithFormat.Save(filename)
```

Comment:
nit: again, this it not a biggie, but it feels a bit weird, having to 'clone' the producer just to change the format... and clone is only used for this.

Reply:
An initial implementation that I had was to have a separate type to do the writing / saving and to construct this when `Save` is called. Let me look at it a bit again and see if I can come up with a better abstraction.